### DISCUSSION TITLE
How to read mesh data from model?

### DISCUSSION THREAD

**Discussion by ravengrim** [stride3d/stride#1907] [Q&A] [answered]
I'm trying to get vertex, triangle and normal data from a model.
In Unity I can simply get a copy of the buffers like this:
```
Mesh mesh = GetComponent<MeshFilter>.mesh;
Vector3[] vertices = mesh.vertices;
Vector3[] normals = mesh.normals;
int[] triangles = mesh.triangles;
```

How would I do this in Stride?

**Comment by ravengrim** [comment]
I found a workaround. Instead of trying to read the mesh data from models, I imported my model files as raw assets, referenced them by UrlReference and used the AssimpNet package to read the files like this:
```
AssimpContext importer = new AssimpContext();
Stream stream = Content.OpenAsStream(urlReference.Url, StreamFlags.None);
Assimp.Scene sc = importer.ImportFileFromStream(stream, "gltf");
```

**Comment by Bodyclock-Games** [comment, marked as answer]
New here also from Unity. As far as I know it's something like:

`int[] indices = model.Meshes[0].Draw.IndexBuffer.Buffer.GetData<int>(Game.GraphicsContext.CommandList);`

**Reply by ravengrim** [reply]
I had discovered the GetData<T> method as well but didn't know what I'm supposed to pass in for the commandList parameter. Thanks a lot!

**Comment by lucas-0liveira** [comment]
In Stride, you can achieve similar results by accessing the mesh data through the Mesh class. Here's the equivalent code to obtain vertex, triangle, and normal data:

// Assuming you have a reference to the mesh
Mesh mesh = yourMeshReference;

// Get vertices
Vector3[] vertices = mesh.Vertices.Select(v => v.Position).ToArray();

// Get normals
Vector3[] normals = mesh.Vertices.Select(v => v.Normal).ToArray();

// Get triangles
int[] triangles = mesh.Indices.Select(i => (int)i).ToArray();

**Reply by Eideren** [reply]
Are you a bot ? Answer yes if you aren't

**Reply by lucas-0liveira** [reply]
no

**Comment by Bodyclock-Games** [comment]
I must be missing something because I can't see those properties on the procedural mesh object I created. Do they only exist on mesh assets?
